app/models/concerns/orchestration/dhcp.rb

```diff
@@ -19,58 +19,61 @@ def dhcp?
         !subnet.nil? && subnet.dhcp? && SETTINGS[:unattended] && (!provision? || operatingsystem.present?)
   end
 
-  def dhcp_record
-    return unless dhcp? || @dhcp_record
-    handle_validation_errors do
-      @dhcp_record ||= (provision? && jumpstart?) ? Net::DHCP::SparcRecord.new(dhcp_attrs) : Net::DHCP::Record.new(dhcp_attrs)
+  def dhcp_records
+    return {} unless dhcp?
+    @dhcp_records ||= mac_addresses_for_provisioning.map do |record_mac|
+      build_dhcp_record(record_mac)
     end
   end
 
   def reset_dhcp_record_cache
-    @dhcp_record = nil
+    @dhcp_records = nil
   end
 
   def rebuild_dhcp
-    if dhcp?
-      del_dhcp_safe
-      begin
-        set_dhcp
-      rescue => e
-        Foreman::Logging.exception "Failed to rebuild DHCP record for #{name}, #{ip}", e, :level => :error
-        false
-      end
-    else
+    unless dhcp?
       logger.info "DHCP not supported for #{name}, #{ip}, skipping orchestration rebuild"
-      true
+      return true
+    end
+
+    del_dhcp_safe
+    begin
+      set_dhcp
+    rescue => e
+      Foreman::Logging.exception "Failed to rebuild DHCP record for #{name}, #{ip}", e, :level => :error
+      false
     end
   end
 
   protected
 
   def del_dhcp_safe
-    if dhcp_record
-      begin
-        del_dhcp
-      rescue => e
-        Foreman::Logging.exception "Proxy failed to delete DHCP record for #{name}, #{ip}", e, :level => :error
-      end
+    return unless dhcp_records.any?
+    begin
+      del_dhcp
+    rescue => e
+      Foreman::Logging.exception "Proxy failed to delete DHCP record for #{name}, #{ip}", e, :level => :error
     end
   end
 
   def set_dhcp
-    dhcp_record.create
+    dhcp_records.all? { |record| record.create }
   end
 
   def set_dhcp_conflicts
-    dhcp_record.conflicts.each{|conflict| conflict.create}
+    dhcp_records.all? do |record|
+      record.conflicts.each { |conflict| conflict.create }
+    end
   end
 
   def del_dhcp
-    dhcp_record.destroy
+    dhcp_records.all? { |record| record.destroy }
   end
 
   def del_dhcp_conflicts
-    dhcp_record.conflicts.each{|conflict| conflict.destroy}
+    dhcp_records.all? do |record|
+      record.conflicts.each { |conflict| conflict.destroy }
```

Comment:
These are the possibilities:

No conflicts - `record.conflicts.each` will return [], since it's not `false`, it's `true` for `all?`. This works fine.
Conflicts - `record.conflicts.each` will return [false, true]. That will be counted as `true` for `all?`. This does not work fine. For example:

```ruby
[1,2,3].all? { |e| [e % 2 == 0, true] }
=> true
[1,2,3].all? { |e| e % 2 == 0 }
=> false
```

You should do `record.conflicts.map(&:destroy).uniq == ([true] || [])` to ensure all conflicts were destroyed properly (or there were none to begin with).